powersimdata/input/expansion_candidates.py

```diff
@@ -0,0 +1,146 @@
+from dataclasses import dataclass
+
+import pandas as pd
+
+
+def check_bus_id(bus_id, grid):
+    """Check that buses are within the given grid
+
+    :param pd.Series bus_id: the bus ids to check
+    :param powersimdata.input.grid.Grid grid: reference grid
+    :raises ValueError: if any buses are not in the grid
+    """
+    valid = bus_id.isin(grid.bus.index)
+    if not valid.all():
+        msg = f"Invalid bus id = {list(bus_id[~valid])}"
+        raise ValueError(msg)
+
+
+def check_branch_voltage(branch, grid):
+    """Check that branches are attached to buses with the same voltage
+
+    :param pd.DataFrame branch: dataframe with from_bus and to_bus columns
+    :param powersimdata.input.grid.Grid grid: reference grid
+    :raises ValueError: if any branches have mismatched voltage
+    """
+    basekv = grid.bus.loc[:, "baseKV"]
+    v1 = basekv[branch["from_bus"]].reset_index(drop=True)
+    v2 = basekv[branch["to_bus"]].reset_index(drop=True)
+    cmp = v1.compare(v2)
+    if not cmp.empty:
+        mismatch = list(cmp.index)
+        raise ValueError(
+            f"from_bus and to_bus must have the same baseKV. rows={mismatch}"
+        )
+
+
+def check_branch(branch, grid):
+    """Check branch expansion candidates relative to a specific grid
+
+    :param pd.DataFrame branch: dataframe of branch candidates
+    :param powersimdata.input.grid.Grid grid: reference grid
+    """
+    check_bus_id(branch.from_bus, grid)
+    check_bus_id(branch.to_bus, grid)
+    check_branch_voltage(branch, grid)
+
+
+def check_plant(plant, grid):
+    """Check plant expansion candidates relative to a specific grid
+
+    :param pd.DataFrame plant: dataframe of plant candidates
+    :param powersimdata.input.grid.Grid grid: reference grid
+    """
+    check_bus_id(plant.bus_id, grid)
+
+
+def check_storage(storage, grid):
+    """Check storage expansion candidates relative to a specific grid
+
+    :param pd.DataFrame storage: dataframe of storage candidates
+    :param powersimdata.input.grid.Grid grid: reference grid
+    """
+    check_bus_id(storage.bus_id, grid)
```

Comment:
It seems that all these functions could be used in a PCM workflow and hence could live in the `powersimdata.input.check` module. To illustrate, we could use these functions when adding new storage, branches, plants and storage in the change table.

Reply:
I had the same thought but didn't want to expand the scope. The bus id check could be easy to refactor. I just noticed we have a function `_check_line_voltages` already but that is used to check the consistency of an existing grid, so we would have to refactor it (or add a separate function) to compare new branches with a given bus dataframe. I'm not seeing a check for that in the change table - is that something we should have? If it's missing, that might be worth adding, but I was planning to hold off on refactoring until there is a bit more cem progress.

Reply:
Sounds good.

I believe we check that the bus(es) exist when adding a new line, a plant or a storage in the change table. You are right, we don't check the voltages. It could be added.